### ISSUE TITLE
Reproducibility of task destination in parallel

### ISSUE THREAD

**Issue by EvanKomp** [ipython/ipyparallel#420]
Context: I am running scikit-learn Random search in parallel, and I would like it to be reproducible down to weight initialization, etc. This is easily done on one CPU. Attempting this with Joblib ipyparallel backend leads to non-reproducible results. Using SLURM cluster.

I could not figure out a way to set some sort of global seed, so my cheap solution was to scatter unique and reproducible seeds to each engine. While each engine gets the seeds, with LoadBalancedView there is no way to ensure that some task ends up on the same engine # everytime the job is run that I can see, so the tasks have a semi-random destination and the resulting grid search is not reproducible.

Is there anyway to control/reproduce which engine gets which task?
Alternatively does anyone know to set a 'global' seed? I assume this would involve communication and overhead.

Appreciation in advance, and let me know if this is not the right place/where else I should go.

The code I am using to start the engines:
```
#create a new ipython profile appended with the job id number
profile=job_${SLURM_JOB_ID}

echo "Creating profile_${profile}"
ipython profile create ${profile}

ipcontroller --ip="*" --profile=${profile} --scheme=lru &
sleep 10

#srun: runs ipengine on each available core
srun ipengine --profile=${profile} --location=$(hostname) &
sleep 25

echo "Launching job for script $1"
python $1 -p ${profile} 1>run.out 2>run.err 
```
And enacted on these engines to run the search:
```
from joblib import Parallel, parallel_backend
from joblib import register_parallel_backend
from joblib import delayed
from joblib import cpu_count
from ipyparallel import Client
from ipyparallel.joblib import IPythonParallelBackend
import numpy as np
import pandas as pd
import os
import sys
import argparse
import logging

from sklearn.model_selection import KFold
from numpy.random import seed
import tensorflow as tf

import modeling_utils_36 as mlt
seed(1)
tf.random.set_seed(1)
FILE_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append(FILE_DIR)
#prepare the logger
parser = argparse.ArgumentParser()
parser.add_argument("-p", "--profile", default="ipy_profile",
                 help="Name of IPython profile to use")
args = parser.parse_args()
profile = args.profile
logging.basicConfig(filename=os.path.join(FILE_DIR,profile+'.log'),
                    filemode='w',
                    level=logging.DEBUG)
logging.info("number of CPUs found: {0}".format(cpu_count()))
logging.info("args.profile: {0}".format(profile))

#prepare the engines
c = Client(profile=profile)
#The following command will make sure that each engine is running in
# the right working directory to access the custom function(s).
c[:].map(os.chdir, [FILE_DIR]*len(c))
logging.info("c.ids :{0}".format(str(c.ids)))
bv = c.load_balanced_view()
dv = c.direct_view()

register_parallel_backend('ipyparallel',
                          lambda : IPythonParallelBackend(view=bv))

## set seeds on each core
# send the individual seeds
seeds = list(range(len(c)))
dv.scatter('seed_no', seeds)
#execute block of code
ex_code = '''
from numpy.random import seed
import tensorflow as tf
#import modeling_utils_35 as mlt
seed(seed_no[0])
tf.random.set_seed(seed_no[0])
'''
dv.execute(ex_code, silent=False, block=True)

print(dv['seed_no'])

dev_data = pd.read_csv('XXX')

#cv = KFold(3, shuffle=True, random_state=1)
ind = KFold(3).split(dev_data)
ind_list = [i for i in ind]
take = ind_list[0]
cv = [[take, take, take]]

space = {'epochs': 5, 'n_iter':14, 'learning_rate': (.01, .1), 'verbose':10, 'n_jobs':-1, 'cv': cv}

with parallel_backend('ipyparallel'):
        seed(1)
        tf.random.set_seed(1)
        results, hyp_cache = mlt.random_search_KERAS(dev_data, 'l', mlt.KERAS_regressor_dense, space)

print(c.history)
```

**Comment by minrk** [comment]
I don't think that random seeds are the issue here, rather the randomness inherent in races in load-balanced tasks, since tasks are assigned to engines on an as-available basis, and nondeterministic timing issues in resource allocation, network, etc. can result in different engines completing their tasks in a different order on each run.

You can make the graph deterministic by assigning engines explicitly in the joblib Backend, rather than relying on the LoadbalancedView:

```python
from joblib import register_parallel_backend
from ipyparallel.joblib import register, IPythonParallelBackend


class DeterministicBackend(IPythonParallelBackend):
    """IPython Parallel joblib backend with deterministic engine assignment
    
    assigns tasks in round-robin order instead of using a load-balanced view
    Engine assignment is deterministic *for a given number of engines*.
    """
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._task_counter = 0
        self.client = self._view.client
    
    def apply_async(self, *args, **kwargs):
        # assign tasks round-robin-style randomly via dire
        # use random assignment instead of load-balancing
        # to use local random state for deterministic graph
        self._task_counter = (self._task_counter + 1) % len(self.client)
        self._view = self.client[self.client.ids[self._task_counter]]
        return super().apply_async(*args, **kwargs)
    

register_parallel_backend("ipyparallel-deterministic", DeterministicBackend)
```

Of course, this actually disables any load balancing, but it should do okay as long as there are enough tasks and they are uniform-enough in their cost. You could also similarly record the history of engine assignment for a given run with the real load-balancer and use that for the sequence of targets to consume in the backend for future runs, but that's probably more information to keep track of than you really want.

The way to make it truly deterministic, though, is to ensure that *each task* is deterministic, which means either consuming all random state in the construction of the tasks (no random API calls within tasks), or including the explicit setting of random state as a local operation *of each task*, so that it's deterministic no matter where it runs. I'm not sure exactly how the tasks are being constructed to say how easy it would be to accomplish that. If you did construct tasks in this way, it would solve this determinism problem for *all* joblib backends, not just IPython Parallel, so that might be worth investigating.
